tendermint/src/merkle.rs

```diff
@@ -1,49 +1,49 @@
 //! Merkle tree used in Tendermint networks
 
 pub mod proof;
-
 pub use proof::Proof;
 
-use core::marker::PhantomData;
-
-use digest::{consts::U32, Digest, FixedOutputReset};
-
-use crate::crypto::Sha256;
-use crate::prelude::*;
-
 /// Size of Merkle root hash
-pub use crate::crypto::sha256::HASH_SIZE;
+pub use crate::crypto::HASH_SIZE;
+use crate::{
+    crypto::{DigestSha256, HashType},
+    prelude::*,
+};
 
 /// Hash is the output of the cryptographic digest function
 pub type Hash = [u8; HASH_SIZE];
 
-/// Compute a simple Merkle root from vectors of arbitrary byte vectors.
-/// The leaves of the tree are the bytes of the given byte vectors in
-/// the given order.
-pub fn simple_hash_from_byte_vectors<H>(byte_vecs: &[Vec<u8>]) -> Hash
-where
-    H: MerkleHash + Default,
-{
-    let mut hasher = H::default();
-    hasher.hash_byte_vectors(byte_vecs)
-}
-
 /// Implementation of Merkle tree hashing for Tendermint.
-pub trait MerkleHash {
+pub trait MerkleHash: DigestSha256 {
     // tmhash({})
     // Pre and post-conditions: the hasher is in the reset state
     // before and after calling this function.
-    fn empty_hash(&mut self) -> Hash;
+    fn empty_hash(&mut self) -> Hash {
+        // Get the output of an empty digest state.
+        Self::digest_sha256(&Vec::with_capacity(0))
+    }
```

Comment:
Why bake the implementation into the trait?
This pattern is used elsewhere, for example, in `VerificationPredicates`, but I don't think it's needed here.